src/modules/clustering/hierarchical_clustering_method.cpp

```diff
@@ -104,11 +108,31 @@ int junction_distance(Junction const & lhs, Junction const & rhs)
     }
 }
 
+std::vector<Junction> subsample_partition(std::vector<Junction> const & partition, uint16_t const sample_size)
+{
+    assert(partition.size() >= sample_size);
+    std::vector<uint32_t> indices(partition.size());
+    std::iota(indices.begin(), indices.end(), 0);
+    std::random_device rd;
+    std::mt19937 g(rd());
+    std::shuffle(indices.begin(), indices.end(), g);
+    indices.resize(sample_size);
+    std::vector<Junction> subsample{};
+    for (uint32_t index : indices)
+    {
+        subsample.push_back(partition[index]);
+    }
+    return subsample;
```

Comment:
```suggestion
    std::vector<Junction> subsample{};
    std::sample(partition.begin(), partition.end(), std::back_inserter(subsample),
                sample_size, std::mt19937{std::random_device{}()});
    return subsample;
```

Is this not the same thing?

Reply:
Intressting, I thought std::sample is c++20, but I am wrong, it is c++17!
@eldariont you should totally go with @joshuak94  solution. It is short and clean and just a very good solution. :+1:

Reply:
Also wondering if it's worth it to make the function inline. @SGSSGene what do you think?

Reply:
Great suggestion, this is so much simpler and shorter 👍

Reply:
> Also wondering if it's worth it to make the function inline. @SGSSGene what do you think?

No, `inline` is not needed. The `inline` keyword isn't really related to the optimization technique inlining anymore (even though it might originated from there and is still mentioned in the standard).

Any function that could be inlined, might get inlined independently of the keyword `inline`. All functions that are members of classes automatically have the `inline` keyword set. The `inline` keyword actually allows you to break the "one definition rule"

In this case it might(very unlikely) actually do more harm than good.
Lets say you add another .cpp file and accidentally also define the function with the name and signature `subsample_partition(std::vector<Junction> const &, uint16_t)` inside. If no inline is given, you will get a multiple definition error while linking. If both are declared inline, the linker assumes they are the same function and will delete one of them.

Here you can see the same question+answer on Stackoverflow (careful, the wrong answer is the accepted one)
https://stackoverflow.com/a/157929/1293111